src/Microsoft.AspNetCore.Razor.Evolution/RazorSourceDocument.cs

```diff
@@ -39,11 +49,23 @@ private static RazorSourceDocument ReadFromInternal(Stream stream, string filena
             var memoryStream = new MemoryStream();
             stream.CopyTo(memoryStream);
 
-            return new DefaultRazorSourceDocument(memoryStream, encoding, filename);
-        }
+            memoryStream.Seek(0L, SeekOrigin.Begin);
 
-        public abstract string Filename { get; }
+            var reader = encoding == null
+                ? new StreamReader(memoryStream, detectEncodingFromByteOrderMarks: true)
+                : new StreamReader(memoryStream, encoding);
```

Comment:
There's no need for the memory stream here, you can read directly from the provided stream as long as you don't close it.